thinc/legacy/loss.py

```diff
@@ -0,0 +1,266 @@
+from typing import Optional, Sequence, Dict, Union, Tuple
+from typing import cast, List
+from ..types import Floats2d, Ints1d
+from ..config import registry
+from ..util import to_categorical, get_array_module
+from ..loss import CategoricalCrossentropy, IntsOrFloatsOrStrs, Loss
+from ..loss import _make_mask, _make_mask_by_value
+from ..loss import SequenceCategoricalCrossentropy
+
+
+class LegacyCategoricalCrossentropy(Loss):
+    names: Optional[Sequence[str]]
+    missing_value: Optional[Union[str, int]]
+    _name_to_i: Dict[str, int]
+
+    def __init__(
+        self,
+        *,
+        normalize: bool = True,
+        names: Optional[Sequence[str]] = None,
+        missing_value: Optional[Union[str, int]] = None,
+        neg_prefix: Optional[str] = None,
+        label_smoothing: float = 0.0,
+    ):
+        self.normalize = normalize
+        self.names = names
+        self.missing_value = missing_value
+        self.neg_prefix = neg_prefix
+        self.label_smoothing = label_smoothing
+        if names is not None:
+            self._name_to_i = {name: i for i, name in enumerate(names)}
+        else:
+            self._name_to_i = {}
+
+    def convert_truths(self, truths, guesses: Floats2d) -> Tuple[Floats2d, Floats2d]:
```

Comment:
The typing information for `truths` was left out originally I think because it was a bit tricky to get `mypy` to typecheck with some many different possibilities. But I'll try to add it, i think the correct type in the `legacy` loss is `Union[List[int], List[str], Ints1d, Floats2d]` for `truths`.

Reply:
added the type to `truths` along with a small forest of `cast`s to type-check with `mypy`.